File: algorithms/0914.x-of-a-kind-in-a-deck-of-cards/solution.rs
use std::collections::HashMap;

impl Solution {
    pub fn has_groups_size_x(deck: Vec<i32>) -> bool {
        fn gcd(x: i32, y: i32) -> i32 {
            if y == 0 {
                x
            } else {
                gcd(y, x % y)
            }
        }

        let mut map: HashMap<i32, i32> = HashMap::new();
        for d in deck {
            let entry = map.entry(d).or_insert(0);
            *entry += 1;
        }

        let values = map.values().cloned().collect::<Vec<i32>>();
        let mut x = values[0];
        for v in values {
            x = gcd(x, v);
        }

        x >= 2
    }
}

File: algorithms/0914.x-of-a-kind-in-a-deck-of-cards/solution.go
func hasGroupsSizeX(deck []int) bool {
	m := map[int]int{}
	for _, d := range deck {
		if _, ok := m[d]; ok {
			m[d] += 1
		} else {
			m[d] = 1
		}
	}

	vs := []int{}
	for _, v := range m {
		vs = append(vs, v)
	}

	x := vs[0]
	for _, v := range vs {
		x = gcd(x, v)
	}

	return x >= 2
}

func gcd(x int, y int) int {
	if y == 0 {
		return x
	} else {
		return gcd(y, x%y)
	}
}
